packages/jupyter-chat/src/components/input/use-chat-commands.tsx

```diff
@@ -0,0 +1,168 @@
+        return;
+      }
+
+      let newCommands: ChatCommand[] = [];
+      for (const provider of providers) {
+        // TODO: optimize performance when this method is truly async
+        try {
+          newCommands = newCommands.concat(
+            await provider.getChatCommands(currentWord)
+          );
+        } catch (e) {
+          console.error(
+            `Error when getting chat commands from command provider '${provider.id}': `,
+            e
+          );
+        }
+      }
+
+      if (newCommands) {
+        setOpen(true);
+      }
+      setCommands(newCommands);
+    }
+
+    getCommands();
+  }, [input, inputRef]);
+
+  return {
+    autocompleteProps: {
+      open,
+      options: commands,
+      getOptionLabel: (command: ChatCommand) => command.name,
+      renderOption: (
+        defaultProps,
+        command: ChatCommand,
+        __: unknown,
+        ___: unknown
+      ) => {
+        const { key, ...listItemProps } = defaultProps;
+        return (
+          <Box key={key} component="li" {...listItemProps}>
+            {command.name}
+          </Box>
+        );
+      },
+      value: null,
+      autoHighlight: true,
+      freeSolo: true,
+      disableClearable: true,
+      onChange: (
+        _: unknown,
+        command: ChatCommand,
+        reason: AutocompleteChangeReason
+      ) => {
+        if (reason !== 'selectOption') {
+          // only call this callback when a command is selected by the user. the
+          // other reasons provided by MUI should never occur, so this check is
+          // mainly for type safety.
+          return;
+        }
+
+        const cursorIndex = inputRef.current?.selectionStart;
+        if (
+          !chatCommandRegistry ||
+          cursorIndex === null ||
+          cursorIndex === undefined
+        ) {
+          return;
+        }
+
+        const currentWord = getCurrentWord(input, cursorIndex);
+        if (!currentWord) {
+          return;
+        }
+
+        const replaceCurrentWordClosure = (newWord: string) => {
+          replaceCurrentWord(input, cursorIndex, newWord, setInput);
+        };
+
+        chatCommandRegistry.handleChatCommand(
+          command,
+          currentWord,
+          replaceCurrentWordClosure
+        );
+      },
+      onHighlightChange:
+        /**
+         * On highlight change: set `highlighted` to whether an option is
+         * highlighted by the user.
+         *
+         * This isn't called when an option is selected for some reason, so we
+         * need to call `setHighlighted(false)` in `onClose()`.
+         */
+        (_, highlightedOption) => {
+          setHighlighted(!!highlightedOption);
+        },
+      onClose:
+        /**
+         * On close: set `highlighted` to `false` and close the popup by
+         * setting `open` to `false`.
+         */
+        () => {
+          setHighlighted(false);
+          setOpen(false);
+        }
+    },
+    menu: {
+      open,
```

Comment:
The `open` state is returned 2 times, in the `autocompleteProps` and in the `menu` props. Don't know if this expected, but both are used in the input component.

Reply:
I duplicated `open` because `chatCommands.menu.open` is more readable than `chatCommands.autocompleteProps.open`. I also found it helpful to distinguish the props being passed to the `Autocomplete` component (under `autocompleteProps`) from the state we are exposing to `ChatInput` (under `menu`).

We are still using a single source of truth (the `open` state variable), so I don't see a concern here. However, feel free to push changes to this branch if you think differently.